light_io function improvements

cmcqueen opened the pull request:
* seek function for light_io_file.
* flush function for light_io_zstd writes.
* Add new tell functions — for light_io_file, light_io_zlib and light_io_zstd.
* Add a new API call `light_pcapng_tell()` to call the light_io tell functions.

The tell functions would be useful to find the size of the pcapng file when writing. This would allow eg output that stops when a maximum size is reached, or log file rotation.

cmcqueen committed:
light_io_file: Add light_file_seek() to the function pointers

That is, set the fn_seek function pointer to light_file_seek.

cmcqueen committed:
light_io_zstd: Add light_zstd_flush_w() function to flush writes

Set fn_flush to point to light_zstd_flush_w.

cmcqueen committed:
light_io: Add tell functions

The tell functions provide the current offset in the file.

kayoub5 commented on include/light_io.h:
```diff
@@ -34,6 +34,7 @@ LIGHT_API size_t LIGHT_API_CALL light_io_read(light_file fd, void* buf, size_t c
 LIGHT_API size_t LIGHT_API_CALL light_io_write(light_file fd, const void* buf, size_t count);
 
 LIGHT_API int LIGHT_API_CALL light_io_seek(light_file fd, long int offset, int origin);
+LIGHT_API long LIGHT_API_CALL light_io_tell(light_file fd);
```
Using long will limit file size to 2GB, this will be a problem, since PCAPs can typically get bigger than this.

cmcqueen replied:
That's true. I thought of that. But I did it because `light_io_seek()` has parameter `long int offset` so I thought I should use `long` for consistency with that. I assume you would want `light_io_seek()` improved at the same time. Or would you want to preserve its existing API for legacy and create a new function with better types?

For the tell functions, what would you like instead? `ftello()` uses `off_t`, and [the man page for it](https://man7.org/linux/man-pages/man3/ftello.3.html) says

> On some architectures, both `off_t` and `long` are 32-bit types, but defining `_FILE_OFFSET_BITS` with the value 64 (before  including _any_ header  files) will turn `off_t` into a 64-bit type.

I assume you want a solution that works well cross-platform.

Looking at the [zstd source code in `fileio_common.h`](https://github.com/facebook/zstd/blob/d462f691ba6a53bd17de492656af7878c73288c8/programs/fileio_common.h#L77), I see it has a fairly complicated definition of `LONG_SEEK` and `LONG_TELL` that aims to deal with this cross-platform.

kayoub5 replied:
I would say make everything use 64bit, I am not too worried about backward compatibility of light_io_seek since it's not used yet (it was planned to be used in #17 to seek back after reading the magic number, but that never took off)

cmcqueen replied:
What type would you like? I looked at `gzoffset64()` as a reference. It uses `z_off64_t` which seems more complicated than I expected. I guess the goals are:

* 64-bit signed.
* Cross-platform.
* Plays nicely with `FILE *`, zlib and zstd.

kayoub5 replied:
we don't have to worry about backward compatibility here, you can simply use int64_t

cmcqueen replied:
I've committed an update that I hope is suitable. Please let me know if I should change anything further.

kayoub5 commented on src/light_io_zstd.c:
```diff
@@ -214,6 +214,20 @@ size_t light_zstd_write(void* context, const void* buf, size_t count)
+{
+	struct zstd_compression_t* compression = context;
+
+	return ftell(compression->file);
```
This will return the position inside original file, not within compressed stream.

cmcqueen replied:
That was deliberate. My interest is specifically in knowing the actual destination file size when writing files.

The same is true for `light_zlib_tell()`, in which I used `gzoffset64()` rather than `gztell64()`.

kayoub5 replied:
the expectation from an end user, is that whatever number tell gives back, can be used to seek, and that tell function should behave like ftell as much as possible.

having a tell function call gzoffset function instead gztell, is kind of counter intuitive.

cmcqueen replied:
I am also an end user. But I guess I have a different use-case in mind.

I'm still happy to contribute to this. But, to progress on this, I'd want to know:

1. Can you please elaborate on the use-case you have in mind, and summarise the API requirements needed for that use-case?
    If I understand correctly, you want seek and tell functions to operate on the uncompressed data stream, for both read and write. Is that right?
3. I'm interested in this use-case: Find the size of the output pcapng file when writing; this would allow eg output that stops when a maximum size is reached, or log file rotation. Could you please recommend a more reasonable API that I could implement that supports this use-case, without being confusing to people who are interested in the use-case of (1)?

kayoub5 replied:
The topic is a naming issue, if you name the function that you plan to add `offset` instead of `tell`, and document it as such (add comment to explain the function), there would be less confusion about the topic.

Look at the docs of gztell and gzoffset as reference:

https://github.com/madler/zlib/blob/5a82f71ed1dfc0bec044d9702463dbdf84ea3b71/zlib.h#L1608

cmcqueen replied:
I've committed an update that I hope is suitable. I've renamed the `_tell` functions to `_offset`. Is that okay? Please let me know if I should change anything further.

I'd be happy to add `_tell` functions also, in principle, but I don't know how to implement tell or seek for zstd.

kayoub5 commented:
Please add unit tests for the added function

cmcqueen committed:
light_io: Changes due to PR review feedback for tell functions

* Rename functions from "tell" to "offset".
* Make offsets int64_t for "offset" functions.
* Similarly, make offsets int64_t for "seek" functions.
* Add seek and offset functions for light_io_mem. The seek function
  existed, but hadn't been linked to the function pointer.

cmcqueen commented:
> Please add unit tests for the added function

I haven't done this yet, but I will aim to do so soon.

kayoub5 commented on include/light_io.h:
```diff
@@ -33,7 +34,8 @@ LIGHT_API light_file LIGHT_API_CALL light_io_open(const char* file_name, const c
 
-LIGHT_API int LIGHT_API_CALL light_io_seek(light_file fd, long int offset, int origin);
+LIGHT_API int64_t LIGHT_API_CALL light_io_seek(light_file fd, int64_t offset, int origin);
+LIGHT_API int64_t LIGHT_API_CALL light_io_offset(light_file fd);
```
document the function, to avoid mis-use

kayoub5 commented on include/light_pcapng_ext.h:
```diff
@@ -93,6 +93,8 @@ LIGHT_API int LIGHT_API_CALL light_pcapng_close(light_pcapng pcapng);
 
 LIGHT_API int LIGHT_API_CALL light_pcapng_flush(light_pcapng pcapng);
 
+LIGHT_API int64_t LIGHT_API_CALL light_pcapng_offset(light_pcapng pcapng);
```
document the function to avoid mis-use

cmcqueen replied:
How and where would you like it documented? Can you point me to existing documentation of any of the other functions to use as a sample/template?

kayoub5 replied:
A normal comment will do, no formal format is chosen at the moment,

See light_read_block as example.

cmcqueen commented:
I see some automated builds have failed, for missing functions such as `ftello64()` and `gzseek64()`.

Now I think, to use functions that are 64-bit friendly in a cross-platform way, it's necessary to avoid such explicit functions as `ftello64()`. Instead:

* Make the seek and offset functions use `off_t` rather than `int64_t` directly.
* Use functions such as `ftello()` and `gzoffset()`.
* These should "just work" on 64-bit platforms.
* Users of 32-bit platforms that are 64-bit capable can set the C define `_FILE_OFFSET_BITS=64`.

I've made a new commit that implements this.

cmcqueen committed:
light_io: Changes due to PR automated build failures

Don't call 64-bit seek/offset functions directly, because this isn't
supported cross-platform. Use `off_t` and functions such as `fseeko()`.

This hopefully should work on most platforms, with the exception of
Windows which likes to do things differently (TBD).
